common/flatpak-instance.c

```diff
@@ -429,6 +429,76 @@ flatpak_instance_new_for_id (const char *id)
   return flatpak_instance_new (dir);
 }
 
+static void
+maybe_clean_tmp (const char *path)
+{
+  glnx_autofd int dfd = openat (AT_FDCWD, path, O_PATH | O_CLOEXEC);
+  glnx_autofd int lock_fd = -1;
+  struct stat statbuf;
+  struct flock l = {
+      .l_type = F_WRLCK,
+      .l_whence = SEEK_SET,
+      .l_start = 0,
+      .l_len = 0,
+  };
+
+  if (dfd < 0)
+    return;
+
+  if (fstat (dfd, &statbuf) < 0)
+    return;
+
+  /* We certainly don't want to delete someone else's directory */
+  if (statbuf.st_uid != geteuid ())
+    return;
+
+  lock_fd = openat (dfd, ".flatpak-tmpdir", O_RDWR | O_CLOEXEC);
```

Comment:
I made the lock file something that indicates "yes, this is a temporary directory created by Flatpak" rather than using the normal `.ref`, as a safety-catch against deleting directories we didn't create.